File: clt_demonstration.py
#!/usr/bin/env python3
"""
Central Limit Theorem Demonstration - Python Implementation
===========================================================

This script demonstrates the Central Limit Theorem using Monte Carlo 
simulations with exponential distributions.

Author: Ahmed Mosad
Course: Statistical Inference DSAI 307
Date: June 2025

Description: Shows how sample means converge to normality as sample size 
increases, regardless of the underlying population distribution.
"""

import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
from scipy import stats
import pandas as pd
from typing import List, Dict, Tuple
import warnings
warnings.filterwarnings('ignore')

# Set style for better-looking plots
plt.style.use('seaborn-v0_8')
sns.set_palette("husl")

# ============================================================================
# CONFIGURATION
# ============================================================================

# Simulation parameters
N_SIMULATIONS = 1500
SAMPLE_SIZES = [10, 50, 100]
EXPONENTIAL_SCALE = 2  # Mean of exponential distribution
POPULATION_MEAN = EXPONENTIAL_SCALE
POPULATION_STD = EXPONENTIAL_SCALE
RANDOM_SEED = 42

# Set random seed for reproducibility
np.random.seed(RANDOM_SEED)

print("=" * 65)
print("CENTRAL LIMIT THEOREM DEMONSTRATION - PYTHON")
print("=" * 65)
print(f"Simulation Parameters:")
print(f"- Number of simulations: {N_SIMULATIONS}")
print(f"- Sample sizes: {SAMPLE_SIZES}")
print(f"- Population mean (μ): {POPULATION_MEAN}")
print(f"- Population std (σ): {POPULATION_STD}")
print(f"- Distribution: Exponential with scale = {EXPONENTIAL_SCALE}")
print("=" * 65)

# ============================================================================
# HELPER FUNCTIONS
# ============================================================================

def generate_sample_means(n_samples: int, sample_size: int, scale: float = 2) -> np.ndarray:
    """
    Generate sample means from exponential distribution.
    
    Parameters:
    -----------
    n_samples : int
        Number of samples to generate
    sample_size : int
        Size of each sample
    scale : float
        Scale parameter (mean) of exponential distribution
    
    Returns:
    --------
    np.ndarray
        Array of sample means
    """
    samples = np.random.exponential(scale, size=(n_samples, sample_size))
    return np.mean(samples, axis=1)


def calculate_theoretical_std(population_std: float, sample_size: int) -> float:
    """
    Calculate theoretical standard deviation of sample means.
    
    Parameters:
    -----------
    population_std : float
        Population standard deviation
    sample_size : int
        Sample size
    
    Returns:
    --------
    float
        Theoretical standard error
    """
    return population_std / np.sqrt(sample_size)


def perform_normality_test(sample_means: np.ndarray, test_size: int = 1000) -> Tuple[float, float]:
    """
    Perform Shapiro-Wilk normality test.
    
    Parameters:
    -----------
    sample_means : np.ndarray
        Array of sample means
    test_size : int
        Maximum size for test
    
    Returns:
    --------
    Tuple[float, float]
        Test statistic and p-value
    """
    test_sample = sample_means[:test_size] if len(sample_means) > test_size else sample_means
    return stats.shapiro(test_sample)

# ============================================================================
# MAIN SIMULATION
# ============================================================================

print("\nRunning simulations...\n")

# Create figure with subplots
fig, axes = plt.subplots(3, 1, figsize=(12, 16))
fig.suptitle('Central Limit Theorem Demonstration\\nSampling Distribution of Means from Exponential Distribution', 
             fontsize=16, fontweight='bold', y=0.98)

# Store results for analysis
results = {}

# Generate and plot for each sample size
for i, n in enumerate(SAMPLE_SIZES):
    print(f"Processing sample size n = {n}...")
    
    # Generate sample means
    sample_means = generate_sample_means(N_SIMULATIONS, n, EXPONENTIAL_SCALE)
    
    # Calculate statistics
    observed_mean = np.mean(sample_means)
    observed_std = np.std(sample_means, ddof=1)
    theoretical_std = calculate_theoretical_std(POPULATION_STD, n)
    
    # Store results
    results[n] = {
        'sample_means': sample_means,
        'observed_mean': observed_mean,
        'observed_std': observed_std,
        'theoretical_std': theoretical_std
    }
    
    # Create histogram
    axes[i].hist(sample_means, bins=40, density=True, alpha=0.7, 
                color='skyblue', edgecolor='darkblue', linewidth=0.5)
    
    # Add observed density curve
    x = np.linspace(sample_means.min(), sample_means.max(), 100)
    kde = stats.gaussian_kde(sample_means)
    axes[i].plot(x, kde(x), 'b-', linewidth=2, label='Observed Distribution')
    
    # Add theoretical normal curve
    theoretical_curve = stats.norm.pdf(x, POPULATION_MEAN, theoretical_std)
    axes[i].plot(x, theoretical_curve, 'r--', linewidth=2, label='Theoretical Normal')
    
    # Add reference lines
    axes[i].axvline(POPULATION_MEAN, color='red', linestyle=':', linewidth=2, 
                   label=f'Population Mean (μ = {POPULATION_MEAN})')
    axes[i].axvline(observed_mean, color='blue', linestyle=':', linewidth=1, 
                   label=f'Sample Mean ({observed_mean:.3f})')
    
    # Formatting
    axes[i].set_title(f'Sample Size n = {n}\\n'
                     f'Mean = {observed_mean:.3f}, Std = {observed_std:.3f} '
                     f'(Theoretical: {theoretical_std:.3f})', fontsize=12)
    axes[i].set_xlabel('Sample Mean')
    axes[i].set_ylabel('Density')
    axes[i].legend()
    axes[i].grid(True, alpha=0.3)

plt.tight_layout()
plt.show()

# ============================================================================
# STATISTICAL SUMMARY
# ============================================================================

print("\\n" + "=" * 65)
print("STATISTICAL SUMMARY")
print("=" * 65)

# Create summary table
summary_data = []
for n in SAMPLE_SIZES:
    result = results[n]
    # Perform normality test
    stat, p_value = perform_normality_test(result['sample_means'])
    
    summary_data.append({
        'Sample Size (n)': n,
        'Observed Mean': f"{result['observed_mean']:.4f}",
        'Observed Std': f"{result['observed_std']:.4f}",
        'Theoretical Std': f"{result['theoretical_std']:.4f}",
        'Error': f"{abs(result['observed_std'] - result['theoretical_std']):.4f}",
        'Shapiro W': f"{stat:.4f}",
        'p-value': f"{p_value:.6f}",
        'Normal?': 'Yes' if p_value > 0.05 else 'No'
    })

summary_df = pd.DataFrame(summary_data)
print(summary_df.to_string(index=False))

print(f"\\nKey Observations:")
print(f"• Population mean (μ) = {POPULATION_MEAN}")
print(f"• All sample means converge to population mean")
print(f"• Standard deviation decreases as sample size increases (∝ 1/√n)")
print(f"• Observed standard deviations closely match theoretical values")

# ============================================================================
# DETAILED NORMALITY ANALYSIS
# ============================================================================

print("\\n" + "=" * 65)
print("NORMALITY TESTS (Shapiro-Wilk)")
print("=" * 65)

for n in SAMPLE_SIZES:
    result = results[n]
    sample_means = result['sample_means']
    
    # Perform normality test
    stat, p_value = perform_normality_test(sample_means)
    
    print(f"Sample Size n = {n:3d}: W = {stat:.4f}, p-value = {p_value:.6f}", end="")
    
    if p_value > 0.05:
        print(" → Distribution appears normal ✓")
    else:
        print(" → Distribution may not be perfectly normal")

print("\\nNote: As sample size increases, the distribution of sample means")
print("becomes more normal, demonstrating the Central Limit Theorem.")

# ============================================================================
# ADVANCED VISUALIZATIONS
# ============================================================================

print("\\n" + "=" * 65)
print("CREATING ADVANCED VISUALIZATIONS")
print("=" * 65)

# Create comparative visualization
fig, axes = plt.subplots(2, 2, figsize=(15, 12))
fig.suptitle('Central Limit Theorem: Comprehensive Analysis', fontsize=16, fontweight='bold')

# Plot 1: Comparison of distributions
ax1 = axes[0, 0]
for n in SAMPLE_SIZES:
    sample_means = results[n]['sample_means']
    ax1.hist(sample_means, bins=30, alpha=0.5, density=True, label=f'n = {n}')
ax1.set_title('Distribution Comparison')
ax1.set_xlabel('Sample Mean')
ax1.set_ylabel('Density')
ax1.legend()
ax1.grid(True, alpha=0.3)

# Plot 2: Convergence of means
ax2 = axes[0, 1]
sample_sizes_extended = range(5, 101, 5)
means_convergence = []
for n in sample_sizes_extended:
    temp_means = generate_sample_means(500, n, EXPONENTIAL_SCALE)
    means_convergence.append(np.mean(temp_means))

ax2.plot(sample_sizes_extended, means_convergence, 'bo-', markersize=4)
ax2.axhline(y=POPULATION_MEAN, color='red', linestyle='--', label=f'Population Mean = {POPULATION_MEAN}')
ax2.set_title('Convergence of Sample Means')
ax2.set_xlabel('Sample Size')
ax2.set_ylabel('Mean of Sample Means')
ax2.legend()
ax2.grid(True, alpha=0.3)

# Plot 3: Standard error comparison
ax3 = axes[1, 0]
sample_sizes_plot = SAMPLE_SIZES
observed_stds = [results[n]['observed_std'] for n in sample_sizes_plot]
theoretical_stds = [results[n]['theoretical_std'] for n in sample_sizes_plot]

x_pos = np.arange(len(sample_sizes_plot))
width = 0.35

ax3.bar(x_pos - width/2, observed_stds, width, label='Observed', alpha=0.7)
ax3.bar(x_pos + width/2, theoretical_stds, width, label='Theoretical', alpha=0.7)
ax3.set_title('Standard Deviation Comparison')
ax3.set_xlabel('Sample Size')
ax3.set_ylabel('Standard Deviation')
ax3.set_xticks(x_pos)
ax3.set_xticklabels(sample_sizes_plot)
ax3.legend()
ax3.grid(True, alpha=0.3)

# Plot 4: Q-Q plots for normality assessment
ax4 = axes[1, 1]
colors = ['red', 'green', 'blue']
for i, n in enumerate(SAMPLE_SIZES):
    sample_means = results[n]['sample_means'][:200]  # Use subset for clarity
    stats.probplot(sample_means, dist="norm", plot=ax4)
    ax4.get_lines()[-2].set_color(colors[i])
    ax4.get_lines()[-1].set_color(colors[i])
    ax4.get_lines()[-2].set_label(f'n = {n}')

ax4.set_title('Q-Q Plots for Normality')
ax4.legend()
ax4.grid(True, alpha=0.3)

plt.tight_layout()
plt.show()

# ============================================================================
# STATISTICAL INSIGHTS
# ============================================================================

print("\\n" + "=" * 65)
print("STATISTICAL INSIGHTS")
print("=" * 65)

# Calculate effect sizes
print("Effect of Sample Size on Variability:")
for i, n in enumerate(SAMPLE_SIZES[:-1]):
    current_std = results[n]['observed_std']
    next_std = results[SAMPLE_SIZES[i+1]]['observed_std']
    reduction = ((current_std - next_std) / current_std) * 100
    print(f"• n = {n} to n = {SAMPLE_SIZES[i+1]}: {reduction:.1f}% reduction in std deviation")

# Theoretical vs observed comparison
print("\\nAccuracy of Theoretical Predictions:")
for n in SAMPLE_SIZES:
    result = results[n]
    accuracy = (1 - abs(result['observed_std'] - result['theoretical_std']) / result['theoretical_std']) * 100
    print(f"• n = {n}: {accuracy:.1f}% accuracy")

# ============================================================================
# CONCLUSIONS
# ============================================================================

print("\\n" + "=" * 65)
print("CONCLUSIONS")
print("=" * 65)
print("This Python implementation demonstrates the Central Limit Theorem:")
print()
print("1. CONVERGENCE TO NORMALITY:")
print("   As sample size increases (n: 10 → 50 → 100), the distribution")
print("   of sample means becomes increasingly normal, despite the")
print("   underlying exponential distribution being highly skewed.")
print()
print("2. MEAN CONVERGENCE:")
print("   All sample means converge to the true population mean (μ = 2),")
print("   regardless of sample size.")
print()
print("3. VARIANCE REDUCTION:")
print("   The standard deviation of sample means decreases as sample")
print("   size increases, following the σ/√n relationship.")
print()
print("4. THEORETICAL VALIDATION:")
print("   Our observed standard deviations closely match theoretical")
print("   predictions, confirming the mathematical foundation of CLT.")
print()
print("PRACTICAL APPLICATIONS:")
print("• Quality control in manufacturing")
print("• Survey research and polling")
print("• Clinical trials and medical research")
print("• Financial risk assessment")
print("• A/B testing in data science")
print()
print("=" * 65)
print("SIMULATION COMPLETE!")
print("=" * 65)

# Save results
print("\\nSaving results...")
# Create outputs directory if it doesn't exist
os.makedirs('outputs', exist_ok=True)

# Convert integer keys to strings for np.savez
results_for_save = {f'n_{k}': v for k, v in results.items()}
np.savez('outputs/clt_simulation_results.npz', **results_for_save)
print("Results saved to 'outputs/clt_simulation_results.npz'")

if __name__ == "__main__":
    print("\\nCentral Limit Theorem demonstration completed successfully!")
    print("Check the generated plots and statistical summaries above.")

File: clt_demonstration.R
# ============================================================================
# Central Limit Theorem Demonstration
# Statistical Inference Project
# ============================================================================
# 
# This script demonstrates the Central Limit Theorem using Monte Carlo 
# simulations with exponential distributions.
# 
# Author: Ahmed Mosad
# Course: Statistical Inference DSAI 307
# Date: June 2025
# 
# Description: Shows how sample means converge to normality as sample size 
# increases, regardless of the underlying population distribution.
# ============================================================================

# Install and load required packages
required_packages <- c("ggplot2", "gridExtra", "dplyr", "grid")

for (package in required_packages) {
  if (!require(package, character.only = TRUE)) {
    install.packages(package, dependencies = TRUE)
    library(package, character.only = TRUE)
  }
}

# Load packages
library(grid)
library(ggplot2)
library(gridExtra)
library(dplyr)

# ============================================================================
# SIMULATION PARAMETERS
# ============================================================================

set.seed(123)
n_simulations <- 1500                    # Number of Monte Carlo simulations
sample_sizes <- c(10, 50, 100)          # Sample sizes to test
exponential_rate <- 1/2                 # Rate parameter (λ = 1/μ)
population_mean <- 1/exponential_rate   # Population mean μ = 2
population_sd <- 1/exponential_rate     # Population std σ = 2

cat("=================================================================\n")
cat("CENTRAL LIMIT THEOREM DEMONSTRATION\n")
cat("=================================================================\n")
cat("Simulation Parameters:\n")
cat("- Number of simulations:", n_simulations, "\n")
cat("- Sample sizes:", paste(sample_sizes, collapse = ", "), "\n")
cat("- Population mean (μ):", population_mean, "\n")
cat("- Population std (σ):", population_sd, "\n")
cat("- Distribution: Exponential with rate λ =", exponential_rate, "\n")
cat("=================================================================\n\n")

# ============================================================================
# HELPER FUNCTIONS
# ============================================================================

#' Generate sample means from exponential distribution
#' 
#' @param n_samples Number of samples to generate
#' @param sample_size Size of each sample
#' @param rate Rate parameter of exponential distribution
#' @return Vector of sample means
generate_sample_means <- function(n_samples, sample_size, rate = 1/2) {
  samples <- matrix(rexp(n_samples * sample_size, rate = rate),
                   nrow = n_samples)
  return(rowMeans(samples))
}

#' Calculate theoretical standard error
#' 
#' @param population_sd Population standard deviation
#' @param sample_size Sample size
#' @return Theoretical standard error
calculate_theoretical_se <- function(population_sd, sample_size) {
  return(population_sd / sqrt(sample_size))
}

# ============================================================================
# MAIN SIMULATION - BASE R IMPLEMENTATION
# ============================================================================

cat("Running simulations...\n\n")

# Set up plotting area (3 rows, 1 column)
par(mfrow = c(3, 1), mar = c(4, 4, 3, 2))

# Store results for summary
results <- list()

# Run simulation for each sample size
for(i in seq_along(sample_sizes)) {
  n <- sample_sizes[i]
  
  cat("Processing sample size n =", n, "...\n")
  
  # Generate exponential samples and calculate means
  samples <- matrix(rexp(n_simulations * n, rate = exponential_rate),
                   nrow = n_simulations)
  sample_means <- rowMeans(samples)
  
  # Calculate statistics
  observed_mean <- mean(sample_means)
  observed_sd <- sd(sample_means)
  theoretical_se <- calculate_theoretical_se(population_sd, n)
  
  # Store results
  results[[i]] <- list(
    sample_size = n,
    observed_mean = observed_mean,
    observed_sd = observed_sd,
    theoretical_se = theoretical_se,
    sample_means = sample_means
  )
  
  # Create histogram
  hist(sample_means,
       prob = TRUE,
       main = paste("Sample Size n =", n),
       sub = paste("Mean =", round(observed_mean, 3), 
                  ", SD =", round(observed_sd, 3),
                  "(Theoretical:", round(theoretical_se, 3), ")"),
       xlab = "Sample Mean",
       ylab = "Density",
       breaks = 40,
       col = "lightblue",
       border = "darkblue",
       cex.main = 1.2,
       cex.sub = 0.9)
  
  # Add density curve for observed data
  lines(density(sample_means), col = "blue", lwd = 2)
  
  # Add theoretical normal curve
  x_range <- seq(min(sample_means), max(sample_means), length = 100)
  theoretical_density <- dnorm(x_range, mean = population_mean, sd = theoretical_se)
  lines(x_range, theoretical_density, col = "red", lwd = 2, lty = 2)
  
  # Add vertical lines
  abline(v = population_mean, col = "red", lty = 3, lwd = 2)
  abline(v = observed_mean, col = "blue", lty = 3, lwd = 1)
  
  # Add legend
  legend("topright", 
         legend = c("Observed Distribution", "Theoretical Normal", "Population Mean", "Sample Mean"),
         col = c("blue", "red", "red", "blue"),
         lty = c(1, 2, 3, 3),
         lwd = c(2, 2, 2, 1),
         cex = 0.8)
}

# Reset plotting parameters
par(mfrow = c(1, 1))

# ============================================================================
# STATISTICAL SUMMARY
# ============================================================================

cat("\n=================================================================\n")
cat("STATISTICAL SUMMARY\n")
cat("=================================================================\n")

# Create summary table
cat(sprintf("%-12s %-12s %-12s %-15s %-10s\n", 
           "Sample Size", "Obs. Mean", "Obs. SD", "Theoretical SE", "Error"))
cat(paste(rep("-", 65), collapse = ""), "\n")

for(i in seq_along(results)) {
  result <- results[[i]]
  error <- abs(result$observed_sd - result$theoretical_se)
  
  cat(sprintf("%-12d %-12.4f %-12.4f %-15.4f %-10.4f\n",
             result$sample_size,
             result$observed_mean,
             result$observed_sd,
             result$theoretical_se,
             error))
}

cat("\nKey Observations:\n")
cat("• Population mean (μ) =", population_mean, "\n")
cat("• All sample means converge to population mean\n")
cat("• Standard deviation decreases as sample size increases (∝ 1/√n)\n")
cat("• Observed SDs closely match theoretical predictions\n")

# ============================================================================
# NORMALITY TESTS
# ============================================================================

cat("\n=================================================================\n")
cat("NORMALITY TESTS (Shapiro-Wilk)\n")
cat("=================================================================\n")

for(i in seq_along(results)) {
  result <- results[[i]]
  n <- result$sample_size
  sample_means <- result$sample_means
  
  # Perform Shapiro-Wilk test (use subset for large samples)
  test_sample <- if(length(sample_means) > 5000) {
    sample(sample_means, 1000)
  } else {
    sample_means
  }
  
  shapiro_result <- shapiro.test(test_sample)
  
  cat(sprintf("Sample size n = %3d: W = %.4f, p-value = %.6f",
             n, shapiro_result$statistic, shapiro_result$p.value))
  
  if(shapiro_result$p.value > 0.05) {
    cat(" → Normal ✓\n")
  } else {
    cat(" → Not perfectly normal\n")
  }
}

# ============================================================================
# ADVANCED VISUALIZATION WITH GGPLOT2
# ============================================================================

cat("\n=================================================================\n")
cat("CREATING ADVANCED VISUALIZATIONS\n")
cat("=================================================================\n")

# Create ggplot2 visualizations
plots_list <- list()

for(i in seq_along(sample_sizes)) {
  n <- sample_sizes[i]
  
  # Use the sample_means from our existing results
  sample_means_data <- results[[i]]$sample_means
  obs_mean <- results[[i]]$observed_mean
  obs_sd <- results[[i]]$observed_sd
  theo_se <- results[[i]]$theoretical_se
  
  # Create data frame
  df <- data.frame(means = sample_means_data)
  
  # Create the plot
  p <- ggplot(df, aes(x = means)) +
    geom_histogram(aes(y = after_stat(density)), bins = 40,
                  fill = "lightblue", color = "darkblue", alpha = 0.7) +
    geom_density(color = "blue", linewidth = 1.2) +
    geom_vline(xintercept = population_mean, color = "red", 
              linetype = "dotted", linewidth = 1.5) +
    geom_vline(xintercept = obs_mean, color = "blue", 
              linetype = "dotted", linewidth = 1) +
    labs(
      title = paste("Sample Size n =", n),
      subtitle = paste("Mean =", round(obs_mean, 3), 
                      ", SD =", round(obs_sd, 3),
                      "(Theoretical:", round(theo_se, 3), ")"),
      x = "Sample Mean",
      y = "Density"
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 14, face = "bold"),
      plot.subtitle = element_text(hjust = 0.5, size = 11),
      panel.grid.minor = element_blank()
    )
  
  # Add theoretical normal curve manually
  x_range <- seq(min(sample_means_data), max(sample_means_data), length.out = 100)
  theoretical_density <- dnorm(x_range, mean = population_mean, sd = theo_se)
  normal_df <- data.frame(x = x_range, y = theoretical_density)
  
  p <- p + geom_line(data = normal_df, aes(x = x, y = y), 
                     color = "red", linetype = "dashed", linewidth = 1.2,
                     inherit.aes = FALSE)
  
  plots_list[[i]] <- p
}

# Arrange plots with error handling
cat("Arranging plots...\n")
tryCatch({
  # Try to create with title
  combined_plot <- grid.arrange(
    grobs = plots_list, 
    ncol = 1,
    top = textGrob("Central Limit Theorem Demonstration\nExponential Distribution (μ=2, σ=2)", 
                  gp = gpar(fontsize = 16, fontface = "bold"))
  )
  cat("✅ Advanced visualizations with title created successfully!\n")
}, error = function(e) {
  # If textGrob fails, create without title
  cat("ℹ️ Creating plots without title...\n")
  combined_plot <- do.call(grid.arrange, c(plots_list, list(ncol = 1)))
  cat("✅ Advanced visualizations created successfully!\n")
})

# Try to save the plot
# Create outputs directory if it doesn't exist
if (!dir.exists("outputs")) {
  dir.create("outputs")
}

tryCatch({
  ggsave("outputs/clt_demonstration_plots.png", 
         plot = last_plot(), 
         width = 12, 
         height = 16, 
         dpi = 300)
  cat("💾 Plots saved as 'outputs/clt_demonstration_plots.png'\n")
}, error = function(e) {
  cat("ℹ️ Could not save plot file, but visualizations were displayed\n")
})

# ============================================================================
# CONCLUSIONS
# ============================================================================

cat("\n=================================================================\n")
cat("CONCLUSIONS\n")
cat("=================================================================\n")
cat("This demonstration clearly shows the Central Limit Theorem:\n\n")

cat("1. CONVERGENCE TO NORMALITY:\n")
cat("   As sample size increases (n: 10 → 50 → 100), the distribution\n")
cat("   of sample means becomes increasingly normal, despite the\n")
cat("   underlying exponential distribution being highly skewed.\n\n")

cat("2. MEAN CONVERGENCE:\n")
cat("   All sample means converge to the true population mean (μ = 2),\n")
cat("   regardless of sample size.\n\n")

cat("3. VARIANCE REDUCTION:\n")
cat("   The standard deviation of sample means decreases as sample\n")
cat("   size increases, following σ/√n relationship.\n\n")

cat("4. THEORETICAL VALIDATION:\n")
cat("   Observed standard deviations closely match theoretical\n")
cat("   predictions, confirming the mathematical foundation.\n\n")

cat("PRACTICAL IMPLICATIONS:\n")
cat("• Statistical inference becomes possible\n")
cat("• Confidence intervals can be constructed\n")
cat("• Hypothesis testing is enabled\n")
cat("• Quality control applications\n\n")

cat("=================================================================\n")
cat("SIMULATION COMPLETE!\n")
cat("=================================================================\n")

# Save workspace for future analysis
save.image("outputs/clt_simulation_results.RData")
cat("Results saved to 'outputs/clt_simulation_results.RData'\n")
